doc/adding_lints.md

```diff
@@ -0,0 +1,366 @@
+This is good, because it makes writing this particular lint less complicated.
+
+We have to make this decision with every new Clippy lint. It boils down to using
+either [`EarlyLintPass`][early_lint_pass] or [`LateLintPass`][late_lint_pass].
+This is a result of Rust's compilation process. You can read more about it [in
+the rustc guide][compilation_stages].
+
+In short, the `LateLintPass` has access to type information while the
+`EarlyLintPass` doesn't. If you don't need access to type information, use the
+`EarlyLintPass`. The `EarlyLintPass` is also faster. However linting speed
+hasn't really been a concern with Clippy so far.
+
+Since we don't need type information for checking the function name, we are
+going to use the `EarlyLintPass`. It has to be imported as well, changing our
+imports to:
+
+```rust
+use rustc::lint::{LintArray, LintPass, EarlyLintPass, EarlyContext};
+use rustc::{declare_tool_lint, lint_array};
+```
+
+### Emitting a lint
+
+With UI tests in place, we can start working on the implementation of the lint logic. We can keep executing the tests until we make them pass.
+
+Let's start by implementing the `EarlyLintPass` for our `FooFunctionsPass`:
+
+```rust
+impl EarlyLintPass for FooFunctionsPass {
+    fn check_fn(&mut self, cx: &EarlyContext<'_>, fn_kind: FnKind<'_>, _: &FnDecl, span: Span, _: NodeId) {
+        // TODO: Emit lint here
+    }
+}
+```
+
+We implement the [`check_fn`][check_fn] method from the
+[`EarlyLintPass`][early_lint_pass] trait. This gives us access to various
+information about the function that is currently being checked. More on that in
+the next section. Let's worry about the details later and emit our lint for
+*every* function definition first.
+
+Depending on how complex we want our lint message to be, we can choose from a
+variety of lint emission functions.  They can all be found in
+[`clippy_lints/src/utils/diagnostics.rs`][diagnostics].
+
+
+```rust
+impl EarlyLintPass for Pass {
+    fn check_fn(&mut self, cx: &EarlyContext<'_>, _: FnKind<'_>, _: &FnDecl, span: Span, _: NodeId) {
+        span_help_and_lint(
+            cx,
+            FOO_FUNCTIONS,
+            span,
+            "function named `foo`",
+            "consider using a more meaningful name"
+        );
+    }
+}
+```
+
+### Adding the lint logic
+
+Writing the logic for your lint will most likely be different from this example,
+so this section is kept rather short.
+
+Using the [`check_fn`][check_fn] method gives us access to [`FnKind`][fn_kind]
+that has two relevant variants for us `FnKind::ItemFn` and `FnKind::Method`.
+Both provide access to the name of the function/method via an [`Ident`][ident].
+
+With that we can expand our `check_fn` method to:
+
+```rust
+impl EarlyLintPass for Pass {
+    fn check_fn(&mut self, cx: &EarlyContext<'_>, fn_kind: FnKind<'_>, _: &FnDecl, span: Span, _: NodeId) {
+        if !is_foo_fn(fn_kind) { return; }
+        span_help_and_lint(
```

Comment:
Bikeshed: Personally I would prefer to write
```rust
if is_foo_fn(fn_kind) {
    span_help_and_lint(..);
}
```
instead of an early return. This is easier to read IMHO. "If is_foo_function then lint" vs "If it's not a foo function then return otherwise lint".